src/chromatix/functional/propagation.py

```diff
@@ -274,32 +253,123 @@ def asm_propagate(
+    propagator: Array,
+    dx: Union[float, Array] = None,
+    N_out: Tuple[int, int] = None,
+    shift_yx: Union[Array, Tuple[float, float]] = (0.0, 0.0),
+    use_czt: bool = False,
+) -> Field:
     """
     Propagate an incoming ``Field`` by the given propagation kernel
     (``propagator``). This amounts to performing a Fourier convolution of the
     ``field`` and the ``propagator``.
     """
     axes = field.spatial_dims
-    u = ifft(fft(field.u, axes=axes) * propagator, axes=axes)
+    if dx is None and N_out is None and not use_czt:
+        u = ifft(fft(field.u, axes=axes) * propagator, axes=axes)
+    else:
+        if N_out is None:
+            N_out = field.spatial_shape
+        if dx is None:
+            dx = field.dx.squeeze()
+        # TODO support vector field?
+        output_field = ScalarField.create(
+            dx=dx,
+            spectrum=field.spectrum,
+            spectral_density=field.spectral_density,
+            shape=N_out,
+            shift_yx=np.array(shift_yx)[..., np.newaxis],
+        )
+        # Scaling factor in Eq 7 of "Band-limited angular spectrum numerical
+        # propagation method with selective scaling of observation window size
+        # and sample number"
+        alpha = output_field.dx / field.dk
+
+        # output field in k-space
+        u = fft(field.u, axes=axes, shift=True) * jnp.fft.fftshift(propagator)
+
+        if use_czt:
+            (y_min, y_max), (x_min, x_max) = output_field.spatial_limits
+            limits_min = [y_min, x_min]
+            limits_max = [y_max, x_max]
+            T = field.surface_area.squeeze()
+
+            # loop over dimensions
+            for d in range(len(axes)):
+                # -- chirp z-transform
+                m = N_out[d]
+                a = jnp.exp(-1j * 2 * jnp.pi / T[d] * limits_min[d])
+                w = jnp.exp(
+                    1j * (2 * jnp.pi / T[d]) * (limits_max[d] - limits_min[d]) / (m - 1)
+                )
+                u = czt(x=u, m=m, a=a, w=w, axis=axes[d])
+
+                # -- modulate
+                N = (m - 1) // 2
+                u = jnp.moveaxis(u, axes[d], -1)
+                C = w ** (-N * jnp.arange(m)) * (a**N)
+                u *= C  # applied to last dimension
+                u = jnp.moveaxis(u, -1, axes[d])
+
+            u *= jnp.prod(1 / alpha)
+
+        else:
+            # Eq 9 of "Band-limited angular spectrum numerical propagation method
+            # with selective scaling of observation window size and sample number"
+            # (2012)
+            wn = alpha * field.k_grid
+            f = jnp.prod(jnp.exp(-1j * jnp.pi / alpha * wn**2), axis=0)
+            B = u * jnp.prod((1 / alpha) * jnp.exp(1j * jnp.pi / alpha * wn**2), axis=0)
+            mod_terms = jnp.prod(
+                output_field.dx * jnp.exp(1j * jnp.pi / alpha * output_field.grid**2),
+                axis=0,
+            )
+            u = mod_terms * fftconvolve(B, f, mode="same", axes=axes)
+
```

Comment:
It looks like the final returned field doesn't get the updated shift.

Reply:
just realized that you probably meant the grid of `field`, not just the diffraction pattern?

Reply:
Yes I meant the grid of the field 👍 So it looks like you are now handling that shifting above.